### PR TITLE
Prefix for negating values from jwt token used in rbaac

### PR DESCRIPTION
**Pull request by Ferdudas97**

### CODE REVIEW

**Comment by KSmigielski** [comment]
I missed some tests (propably integration). For example how it will work if someone set something like that: `"A", "!B"`? It will allow traffic for everyone except `B` and `A` configuration doesn't matter?

**Comment by Ferdudas97** [comment]
> I missed some tests (propably integration). For example how it will work if someone set something like that: `"A", "!B"`? It will allow traffic for everyone except `B` and `A` doesn't matter?

It will accept traffic from users from users which have B and not have A. Negated value works as blackList

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt [hunk 1]
```diff
@@ -112,7 +113,54 @@ internal class RBACFilterFactoryJwtTest : RBACFilterFactoryTestUtils {
...
+        val negatedSelector = "team2"
+        val client = "oauth-prefix"
+        val oAuthPrincipal =
+            oAuthClientPrincipal(getTokenFieldForClientWithSelector("oauth-provider", client), selector)
+        val negatedOAuthPrincipal =
+            oAuthClientPrincipal(getTokenFieldForClientWithSelector("oauth-provider", client), negatedSelector, true)
+        val expectedRbacBuilder = getRBACFilterWithShadowRules(
+            expectedPoliciesForOAuth(
+                withNegatedPrincipal(oAuthPrincipal, negatedOAuthPrincipal),
+                // "$oAuthPrincipal, $negatedOAuthPrincipal",
```
Comment (matb4r) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt:142]: What are this comments for?

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt [hunk 2]
```diff
@@ -257,7 +305,9 @@ internal class RBACFilterFactoryJwtTest : RBACFilterFactoryTestUtils {
     private fun getTokenFieldForClientWithSelector(provider: String, client: String) =
         jwtProperties.providers[provider]!!.matchings[client]!!
 
-    private fun oAuthClientPrincipal(selectorMatching: String, selector: String) = """{
+    private fun oAuthClientPrincipal(selectorMatching: String, selector: String, negated: Boolean = false) =
```
Comment (matb4r) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt:308]: Maybe split to separate functions? Or even just add `"invert": true` conditionally

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt [hunk 3]
```diff
@@ -310,6 +385,13 @@ internal class RBACFilterFactoryJwtTest : RBACFilterFactoryTestUtils {
         }
     """
 
+    private fun withNegatedPrincipal(principal: String, negatedPrincipal: String) = """{
```
Comment (matb4r) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt:388]: `with` does not say anything to me without looking for function usage

**Comment by matb4r** [comment]
Could you give an example how clients should use this feature, to be guest-aware?

**Review by matb4r — CHANGES_REQUESTED** [review]

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt (hunk 3 above)
Comment (Ferdudas97) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt]: changed

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt (hunk 2 above)
Comment (Ferdudas97) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt:306]: changed

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt (hunk 1 above)
Comment (Ferdudas97) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/rbac/RBACFilterFactoryJwtTest.kt]: removed

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 4]
```diff
@@ -356,6 +364,15 @@ class RBACFilterFactory(
         }
     }
 
+    private fun mergePrincipals(principals: Collection<Principal>) = if (principals.size > 1) {
+        Principal.newBuilder()
+            .setAndIds(
+                Principal.Set.newBuilder().addAllIds(principals)
+            ).build()
+    } else {
+        principals.first()
```
[the remark this answers is not in the document]
Comment (Ferdudas97) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt]: changed

**Review by matb4r — APPROVED** [review]

**Review by KSmigielski — APPROVED** [review]

**Pull request merged**